### PR TITLE
Independent tooltip

### PR DESCRIPTION
**Pull request by PavelVanecek**
## Description

Tooltip no longer requires its own specialized method, instead can be rendered together with rest of children.

## Related Issue

https://github.com/recharts/recharts/discussions/3717

## Motivation and Context

I don't want Tooltip to be special cloned element, I want Tooltip to be just another child element so that it can be deeply nested and wrapped in other custom elements.

## How Has This Been Tested?

npm test
storybook

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes

### CODE REVIEW

File: src/chart/SunburstChart.tsx [hunk 1]
```diff
@@ -218,10 +211,9 @@ export const SunburstChart = ({
           role="region"
         >
           <Surface width={width} height={height}>
-            {children}
             <Layer className={layerClass}>{sectors}</Layer>
+            {children}
           </Surface>
-          {renderTooltip()}
```
Comment (PavelVanecek) [inline src/chart/SunburstChart.tsx:224]: Sunburst now renders tooltip only once so I could enable tests for it!

I will fix the other charts too if we find this approach acceptable.

Tooltip still renders everywhere even on this branch.

File: src/component/TooltipBoundingBox.tsx [hunk 2]
```diff
@@ -138,18 +112,21 @@ export class TooltipBoundingBox extends PureComponent<TooltipBoundingBoxProps, S
...
-        }}
-      >
-        {children}
-      </div>
+      <g>
+        <foreignObject x="0" y="0" width="100%" height="100%" style={{ pointerEvents: 'none' }}>
+          {/* This element allow listening to the `Escape` key. // See https://github.com/recharts/recharts/pull/2925 */}
+          <div
+            // @ts-expect-error typescript library does not recognize xmlns attribute
+            xmlns="http://www.w3.org/1999/xhtml"
```
Comment (ckifer) [inline src/component/TooltipBoundingBox.tsx:120]: Why is this necessary?

File: src/component/TooltipBoundingBox.tsx [hunk 3]
```diff
@@ -138,18 +112,21 @@ export class TooltipBoundingBox extends PureComponent<TooltipBoundingBoxProps, S
...
-        className={cssClasses}
-        style={outerStyle}
-        ref={node => {
-          this.wrapperNode = node;
-        }}
-      >
-        {children}
-      </div>
+      <g>
+        <foreignObject x="0" y="0" width="100%" height="100%" style={{ pointerEvents: 'none' }}>
```
Comment (ckifer) [inline src/component/TooltipBoundingBox.tsx:116]: Any downsides of doing this? Doesn't the X and Y matter?

File: src/chart/SunburstChart.tsx (hunk 1 above)
Comment (ckifer) [inline src/chart/SunburstChart.tsx:224]: Does being inside or outside the surface make differences? I guess this is why the foreign object is needed?

**Review by ckifer — COMMENTED** [review]
Just some curiosities, looks great!

File: src/component/TooltipBoundingBox.tsx (hunk 3 above)
Comment (PavelVanecek) [reply src/component/TooltipBoundingBox.tsx:116]: No downside as far as I can tell.

File: src/component/TooltipBoundingBox.tsx (hunk 2 above)
Comment (PavelVanecek) [reply src/component/TooltipBoundingBox.tsx:120]: I added a comment in an attempt to explain

File: src/chart/SunburstChart.tsx (hunk 1 above)
Comment (PavelVanecek) [reply src/chart/SunburstChart.tsx:224]: Yes it does - that's why the `foreignObject` and `xmlns` are necessary.

I suspect that this was the original motivation to have tooltip be a special element rendered outside of the regular children flow. For some reason the original author decided against using `foreignObject` even though it was available since version 1 of each browser? Anyway.

Also, in SunburstChart now (and soon in other charts) - it depends on the order. Tooltip no longer automagically renders above everything, but it now respects the order in its children array. This might be a breaking change for some, or fix for others - I think it's something we should mention in 3.0 changelog as breaking to be on the safe side.

Comment (ckifer) [reply src/chart/SunburstChart.tsx:224]: So you could potentially hide the tooltip behind other elements? That doesn't seem ideal but makes sense I guess

**Comment by ckifer** [comment]
@PavelVanecek  Want to call out that storybook at least throws this

`Warning: <foreignObject /> is using incorrect casing. Use PascalCase for React components, or lowercase for HTML elements.`

when rendering. Not sure if that is always or only in storybook but worth checking on

**Comment by PavelVanecek** [comment]
Yeah it's a warning when this svg element is rendered in HTML context. It's a warning only and it will go away once we finish the refactor.

**Comment by ckifer** [comment]
Cool thanks
